File: src/network/packet/impl/CloudNotificationPacket.php
<?php

namespace pocketcloud\cloud\network\packet\impl;

use pocketcloud\cloud\console\log\CloudLogger;
use pocketcloud\cloud\network\client\ServerClient;
use pocketcloud\cloud\network\packet\ClientboundPacket;
use pocketcloud\cloud\network\packet\CloudboundPacket;
use pocketcloud\cloud\network\packet\CloudPacket;
use pocketcloud\cloud\network\packet\data\NotificationType;
use pocketcloud\cloud\network\packet\util\PacketData;
use pocketcloud\cloud\player\CloudPlayerManager;

final class CloudNotificationPacket extends CloudPacket implements ClientboundPacket, CloudboundPacket {

    public function __construct(
        private ?NotificationType $notificationType = null,
        private array $args = []
    ) {}

    public function handle(ServerClient $client): void {
        $extraArgs = [];
        if ($this->notificationType->canLog()) {
            switch ($this->notificationType) {
                case NotificationType::PLAYER_JOIN_FAILED: {
                    [$player, $server, $reason] = [$this->args["player"], $this->args["server"], $this->args["reason"]];
                    $alreadyOnAServer = CloudPlayerManager::getInstance()->get($player)?->getCurrentServerName() !== null;
                    CloudLogger::get()->info("The player §b{} §rtried to join" . ($alreadyOnAServer ? "" : " via") . " §b{}§r, but got §ckicked§r: §b{}", $player, $server, $this->formatReason($reason));
                    break;
                }
                case NotificationType::PLAYER_KICKED: {
                    [$player, $server, $reason] = [$this->args["player"], $this->args["server"], $this->args["reason"]];
                    CloudLogger::get()->info("The player §b{} §rhas been §ckicked §rfrom §b{}§r: §b{}", $player, $server, $this->formatReason($reason));
                    break;
                }
                default: break;
            }
        }

        $this->notificationType->notify($this->args);
    }

    public function encodePayload(PacketData $packetData): void {
        $packetData->writeAll($this->notificationType, $this->args);
    }

    public function decodePayload(PacketData $packetData): void {
        $packetData->readAllTypeSafe([&$this->notificationType, &$this->args], [fn() => $packetData->readNotificationType(), fn() => $packetData->readArray()]);
    }

    public function getNotificationType(): ?NotificationType {
        return $this->notificationType;
    }

    public function getArgs(): array {
        return $this->args;
    }

    private function formatReason(string $reason): string {
        $newReason = substr(current(explode("\n", $reason)), 0, 100);
        if (strlen($newReason) !== strlen($reason)) $newReason .= "...";
        return $newReason;
    }

    public static function create(NotificationType $notificationType, array $args): self {
        return new self($notificationType, $args);
    }
}

File: cloud-core/src/main/java/de/pocketcloud/cloud/network/packet/impl/CloudNotificationPacket.java
package de.pocketcloud.cloud.network.packet.impl;

import de.pocketcloud.cloud.console.log.CloudLogger;
import de.pocketcloud.cloud.network.client.ServerClient;
import de.pocketcloud.cloud.notification.Notifier;
import de.pocketcloud.network.packet.AuthenticatedPacket;
import de.pocketcloud.network.packet.ClientboundPacket;
import de.pocketcloud.network.packet.CloudboundPacket;
import de.pocketcloud.cloud.network.packet.CloudPacket;
import de.pocketcloud.network.packet.type.NotificationType;
import de.pocketcloud.network.packet.data.PacketData;
import de.pocketcloud.cloud.player.CloudPlayerManager;
import lombok.Getter;
import lombok.NoArgsConstructor;
import org.jetbrains.annotations.NotNull;

import java.util.Map;

@NoArgsConstructor
@Getter
public final class CloudNotificationPacket extends CloudPacket implements ClientboundPacket, CloudboundPacket, AuthenticatedPacket {

    private NotificationType notificationType;
    private Map<String, Object> args;

    public CloudNotificationPacket(NotificationType notificationType, Map<String, Object> args) {
        this.notificationType = notificationType;
        this.args = args != null ? args : Map.of();
    }

    @Override
    public void handle(@NotNull ServerClient client) {
        if (Notifier.canLog(notificationType)) {
            switch (notificationType) {
                case PLAYER_JOIN_FAILED -> {
                    String player = (String) args.get("player");
                    String server = (String) args.get("server");
                    String reason = (String) args.get("reason");
                    boolean alreadyOnAServer = CloudPlayerManager.instance().get(player).filter(p -> p.currentServerName() != null).isPresent();
                    CloudLogger.get().info("The player §b{} §rtried to join" + (alreadyOnAServer ? "" : " via") + " §b{}§r, but got §ckicked§r: §b{}", player, server, formatReason(reason));
                }
                case PLAYER_KICKED -> {
                    String player = (String) args.get("player");
                    String server = (String) args.get("server");
                    String reason = (String) args.get("reason");
                    CloudLogger.get().info("The player §b{} §rhas been §ckicked §rfrom §b{}§r: §b{}", player, server, formatReason(reason));
                }
                default -> {}
            }
        }

        Notifier.notify(notificationType, args, Map.of());
    }

    @Override
    public void encodePayload(PacketData packetData) {
        packetData.writeAll(notificationType, args);
    }

    @Override
    public void decodePayload(PacketData packetData) {
        this.notificationType = packetData.readEnum(NotificationType.class);
        this.args = packetData.readMap();
    }

    private String formatReason(String reason) {
        if (reason == null || reason.isEmpty()) return "No reason applied.";
        String newReason = reason.split("\n")[0];
        if (newReason.length() > 100) newReason = newReason.substring(0, 100);
        if (newReason.length() != reason.length()) newReason += "...";
        return newReason;
    }

    public static CloudNotificationPacket create(NotificationType notificationType, Map<String, Object> args) {
        return new CloudNotificationPacket(notificationType, args);
    }
}
